pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go

```diff
@@ -0,0 +1,361 @@
+			r.logger.WithField("build_name", build.Name).WithField("build_phase", build.Status.Phase).Debug("Storing build")
+			builds[fmt.Sprintf("%s/%s", build.Namespace, build.Name)] = build
+
+			if err := r.updateMultiArchBuildConfig(ctx, ctrlruntimeclient.ObjectKey{Namespace: mabc.Namespace, Name: mabc.Name},
+				func(mabcToMutate *v1.MultiArchBuildConfig) {
+					mabcToMutate.Status.Conditions = append(mabcToMutate.Status.Conditions, constructCondition(build))
+				}); err != nil {
+				return err
+			}
+		}
+
+		if checkAllBuildsFinished(builds) {
+			r.logger.Info("All builds have finished...")
+			return nil
+		}
+	}
+}
+
+func (r *reconciler) createBuildsForArchitectures(ctx context.Context, mabc *v1.MultiArchBuildConfig) (map[string]*buildv1.Build, error) {
+	builds := make(map[string]*buildv1.Build)
+
+	for _, arch := range r.architectures {
+		commonSpec := mabc.Spec.BuildSpec.CommonSpec.DeepCopy()
+		commonSpec.NodeSelector = map[string]string{nodeArchitectureLabel: arch}
+		commonSpec.Output.To.Name = fmt.Sprintf("%s-%s", commonSpec.Output.To.Name, arch)
+
+		build := &buildv1.Build{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", mabc.Name, arch),
+				Namespace: mabc.Namespace,
+				Labels: map[string]string{
+					v1.MultiArchBuildConfigNameLabel: mabc.Name,
+					v1.MultiArchBuildConfigArchLabel: arch,
+				},
+			},
+			Spec: buildv1.BuildSpec{
+				CommonSpec: *commonSpec,
+			},
+		}
+
+		r.logger.WithField("build_namespace", build.Namespace).WithField("build_name", build.Name).Info("Creating build...")
+		if err := r.client.Create(ctx, build); err != nil {
+			return nil, fmt.Errorf("coudldn't create build %s/%s: %w", build.Namespace, build.Name, err)
+		}
+		builds[fmt.Sprintf("%s/%s", build.Namespace, build.Name)] = build
+	}
+	return builds, nil
+}
+
+func (r *reconciler) updateMultiArchBuildConfig(ctx context.Context, namespacedName types.NamespacedName, mutateFn func(mabcToMutate *v1.MultiArchBuildConfig)) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		multiArchBuildConfig := &v1.MultiArchBuildConfig{}
+		if err := r.client.Get(ctx, namespacedName, multiArchBuildConfig); err != nil {
+			return fmt.Errorf("failed to get the MultiArchBuildConfig: %w", err)
+		}
+
+		mutateFn(multiArchBuildConfig)
+
+		r.logger.WithField("namespace", namespacedName.Namespace).WithField("name", namespacedName.Name).Info("Updating MultiArchBuildConfig...")
+		if err := r.client.Update(ctx, multiArchBuildConfig); err != nil {
+			return fmt.Errorf("failed to update MultiArchBuildConfig %s: %w", multiArchBuildConfig.Name, err)
+		}
+		return nil
+	})
+}
+
+func (r *reconciler) pruneBuilds(ctx context.Context, builds map[string]*buildv1.Build) error {
+	for _, build := range builds {
+		if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			b := &buildv1.Build{}
+			if err := r.client.Get(ctx, types.NamespacedName{Namespace: build.Namespace, Name: build.Name}, b); err != nil {
```

Comment:
Why is this `Get` necessary?

Reply:
To avoid conflicts. The resource could change in the meantime by another process that is not related to this controller. This is an idiomatic way to manage the resources in a kube/openshift cluster.

Reply:
But what are we gaining if we're just doing an arbitrary `Get` right before deleting it?  What conflict does this prevent?

If the intent is to delete the specific version created previously, shouldn't we use the values from `builds` when deleting?  Or do it this way:

https://github.com/openshift/ci-tools/blob/2dee011d0b7a2aa21cdf6dce77775a3c9bf4ace2/pkg/util/pods.go#L91

Reply:
In any case,`ci-operator` doesn't delete `Build`s when they are done, we just let the namespace expire (our TTL is very short these days).  This is very useful if someone needs to investigate when something goes wrong.

Reply:
Yea, the whole get/delete thing was to avoid conflicts if the build was mutated in the window between the get and the delete. The delete in that case will fail. However, I would agree that we shouldn't clean them up. I will check your suggestions. Definitely, the users should be able to investigate their build if it fails.